pkg/registry/storage/csinodeinfo/strategy.go

```diff
@@ -0,0 +1,111 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csinodeinfo
+
+import (
+	"context"
+
+	storageapiv1beta1 "k8s.io/api/storage/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/storage"
+	"k8s.io/kubernetes/pkg/apis/storage/validation"
+)
+
+// csiNodeInfoStrategy implements behavior for CSINodeInfo objects
+type csiNodeInfoStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy is the default logic that applies when creating and updating
+// CSINodeInfo objects via the REST API.
+var Strategy = csiNodeInfoStrategy{legacyscheme.Scheme, names.SimpleNameGenerator}
+
+func (csiNodeInfoStrategy) NamespaceScoped() bool {
+	return false
+}
+
+// ResetBeforeCreate clears the Status field which is not allowed to be set by end users on creation.
+func (csiNodeInfoStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+}
+
+func (csiNodeInfoStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	csiNodeInfo := obj.(*storage.CSINodeInfo)
+
+	errs := validation.ValidateCSINodeInfo(csiNodeInfo)
+
+	var groupVersion schema.GroupVersion
+
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
+		groupVersion = schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
+	}
+
+	switch groupVersion {
+	case storageapiv1beta1.SchemeGroupVersion:
+		errs = append(errs, validation.ValidateCSINodeInfo(csiNodeInfo)...)
+	default:
+		// no extra validation
+	}
+	return errs
+}
+
+// Canonicalize normalizes the object after validation.
+func (csiNodeInfoStrategy) Canonicalize(obj runtime.Object) {
+}
+
+func (csiNodeInfoStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// PrepareForUpdate sets the Status fields which is not allowed to be set by an end user updating a CSINodeInfo
+func (csiNodeInfoStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+}
+
+func (csiNodeInfoStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	newCSINodeInfoObj := obj.(*storage.CSINodeInfo)
+	oldCSINodeInfoObj := old.(*storage.CSINodeInfo)
+	errorList := validation.ValidateCSINodeInfo(newCSINodeInfoObj)
```

Comment:
should be `oldObj`?

Reply:
I just followed what is done for VolumeAttachment here: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/registry/storage/volumeattachment/strategy.go#L115
It is the newObj.